### PR TITLE
DCS-52: user notification preferences

### PR DESCRIPTION
**Pull request by bailinhe**
- [x] create DB schema
- [x] implement `CRUD`
- [x] default handling
- [x] add audit events to create and update notification preferences
- [x] clients
- [x] unit-tests
-  API endpoints
    - [x] `POST /notifications/types`
    - [x] `GET /notifications/types`
    - [x] `DELETE /notifications/types/:id`
    - [x] `POST /notifications/targets`
    - [x] `GET /notifications/targets`
    - [x] `DELETE /notifications/targets/:id`
    - [x] `GET /user`
    - [x] `PUT /user`
    - [x] `GET /user/notification-preferences`
    - [x] `PUT /user/notification-preferences`
    - [x] `GET /users/:id`
    - [x] ~~`PUT /users/:id`~~
    - [x] ~~`GET /users/:id/notification-preferences`~~
    - [x] ~~`PUT /users/:id/notification-preferences`~~

### CODE REVIEW

File: pkg/api/v1alpha1/notifications_target.go [hunk 1]
```diff
@@ -0,0 +1,462 @@
...
+
+		return
+	}
+
+	// CRDB cannot refresh a materialized view inside explicit transactions
+	// https://www.cockroachlabs.com/docs/stable/views#known-limitations
+	if err := dbtools.RefreshNotificationDefaults(c.Request.Context(), r.DB); err != nil {
+		sendError(c, http.StatusInternalServerError, err.Error())
+		return
+	}
```
Comment (fishnix) [inline pkg/api/v1alpha1/notifications_target.go:146]: are you calling this because the view wont get updated at all after the transaction is committed? my read of the link is just that you can't do this *inside* a transaction.

File: pkg/api/v1alpha1/notifications_target.go [hunk 2]
```diff
@@ -0,0 +1,462 @@
...
+	if req.Name != "" {
+		sendError(c, http.StatusBadRequest, "modifying notification target name is not allowed")
+		return
+	}
+
+	if req.Description != "" {
+		n.Description = req.Description
+	}
+
+	n.DefaultEnabled = req.DefaultEnabled
```
Comment (fishnix) [inline pkg/api/v1alpha1/notifications_target.go:367]: doesn't this mean the field is always required? what if its true but I don't pass it bc I don't want it to change (like the other fields)?

File: pkg/api/v1alpha1/notifications_types.go [hunk 3]
```diff
@@ -0,0 +1,462 @@
...
+	if req.Name != "" {
+		sendError(c, http.StatusBadRequest, "modifying notification type name is not allowed")
+		return
+	}
+
+	if req.Description != "" {
+		n.Description = req.Description
+	}
+
+	n.DefaultEnabled = req.DefaultEnabled
```
Comment (fishnix) [inline pkg/api/v1alpha1/notifications_types.go:367]: same question here about the expected outcome for omitted values

File: pkg/api/v1alpha1/router.go [hunk 4]
```diff
@@ -438,6 +468,80 @@ func (r *Router) Routes(rg *gin.RouterGroup) {
...
+		r.AuditMW.AuditWithType("CreateNotificationTarget"),
+		r.AuthMW.AuthRequired(createScopesWithOpenID("governor:notifications")),
+		r.mwUserAuthRequired(AuthRoleAdmin),
+		r.createNotificationTarget,
+	)
+
+	rg.PUT(
+		"/notification-targets/:id",
+		r.AuditMW.AuditWithType("UpdateNotificationTarget"),
+		r.AuthMW.AuthRequired(readScopesWithOpenID("governor:notifications")),
```
Comment (fishnix) [inline pkg/api/v1alpha1/router.go:534]: `PUT` should be getting update scopes

File: pkg/api/v1alpha1/router.go [hunk 5]
```diff
@@ -438,6 +468,80 @@ func (r *Router) Routes(rg *gin.RouterGroup) {
...
+		r.updateNotificationTarget,
+	)
+
+	rg.DELETE(
+		"/notification-targets/:id",
+		r.AuditMW.AuditWithType("DeleteNotificationTarget"),
+		r.AuthMW.AuthRequired(createScopesWithOpenID("governor:notifications")),
```
Comment (fishnix) [inline pkg/api/v1alpha1/router.go:541]: `DELETE` should be getting delete scopes

File: pkg/api/v1alpha1/router.go [hunk 6]
```diff
@@ -438,6 +468,80 @@ func (r *Router) Routes(rg *gin.RouterGroup) {
 		r.AuthMW.AuthRequired(readScopesWithOpenID("governor:applications")),
 		r.listApplicationTypeApps,
 	)
+
+	rg.GET(
+		"/notification-types",
+		r.AuditMW.AuditWithType("ListNotificationTypes"),
+		r.AuthMW.AuthRequired(readScopesWithOpenID("governor:notifications")),
```
Comment (fishnix) [inline pkg/api/v1alpha1/router.go:475]: In general, I wonder if `notification-types` and `notification-targets` deserve their own scope types? thoughts?

File: pkg/api/v1alpha1/router.go [hunk 7]
```diff
@@ -119,6 +135,20 @@ func (r *Router) Routes(rg *gin.RouterGroup) {
...
+		"/users/:id/notification-preferences",
+		r.AuditMW.AuditWithType("GetUserNotificationPreferences"),
+		r.AuthMW.AuthRequired(readScopesWithOpenID("governor:users")),
+		r.getUserNotificationPreferences,
+	)
+
+	rg.PUT(
+		"/users/:id/notification-preferences",
+		r.AuditMW.AuditWithType("UpdateUserNotificationPreferences"),
+		r.AuthMW.AuthRequired(readScopesWithOpenID("governor:users")),
```
Comment (fishnix) [inline pkg/api/v1alpha1/router.go:148]: this is an update, not a read

File: pkg/api/v1alpha1/users.go [hunk 8]
```diff
@@ -217,6 +232,19 @@ func (r *Router) createUser(c *gin.Context) {
...
+	updateNotificationPreferencesError := error(nil)
+
+	wg.Add(1)
+
+	updateNotificationPreferences := func() {
+		defer wg.Done()
+
+		_, updateNotificationPreferencesStatusCode, updateNotificationPreferencesError = handleUpdateNotificationPreferencesRequests(c, r.DB, user, r.EventBus, req.NotificationPreferences)
+	}
+	go updateNotificationPreferences()
```
Comment (fishnix) [inline pkg/api/v1alpha1/users.go:246]: does this really need to be async? if these are slow, would it be better to make a separate call?

File: pkg/api/v1alpha1/users.go [hunk 9]
```diff
@@ -375,6 +410,19 @@ func (r *Router) updateUser(c *gin.Context) {
...
```
Comment (fishnix) [inline pkg/api/v1alpha1/users.go:424]: same question as above

File: pkg/events/v1alpha1/events.go [hunk 10]
```diff
@@ -33,16 +33,24 @@ const (
 	GovernorApplicationLinkRequestsEventSubject = "applinks.requests"
 	// GovernorApplicationTypesEventSubject is the subject name for application type events (minus the subject prefix)
 	GovernorApplicationTypesEventSubject = "applicationtypes"
+	// GovernorNotificationTypesEventSubject is the subject name for notification type events (minus the subject prefix)
+	GovernorNotificationTypesEventSubject = "notificationtypes"
+	// GovernorNotificationTargetsEventSubject is the subject name for notification target events (minus the subject prefix)
+	GovernorNotificationTargetsEventSubject = "notificationtarget"
+	// GovernorNotificationPreferencesEventSubject is the subject name for notification preference events (minus the subject prefix)
+	GovernorNotificationPreferencesEventSubject = "notificationpreference"
```
Comment (fishnix) [inline pkg/events/v1alpha1/events.go:41]: maybe these should be `notifications.foo`, then an addon could subscribe to all events about notifications if it wanted

File: sqlboiler.toml [hunk 11]
```diff
@@ -9,4 +9,4 @@ host = "localhost"
 port = 26257
 user = "root"
 sslmode = "disable"
-blacklist = ["goose_db_version"]
+blacklist = ["goose_db_version", "notification_defaults"]
```
Comment (fishnix) [inline sqlboiler.toml:12]: why is this needed?

File: internal/dbtools/notification_preferences.go [hunk 12]
```diff
@@ -0,0 +1,303 @@
...
+		queryMods = append(queryMods, qm.From("notification_defaults as nd"))
+		queryMods = append(queryMods, qm.FullOuterJoin("np on (np.target_id = nd.target_id AND np.type_id = nd.type_id)"))
+		queryMods = append(queryMods, qm.GroupBy("nd.type_slug"))
+	} else {
+		queryMods = append(queryMods, qm.Select("np.type_slug AS notification_type"))
+		queryMods = append(queryMods, qm.Select("jsonb_agg((np.target_slug, np.enabled)) AS notification_targets"))
+		queryMods = append(queryMods, qm.From("np"))
+		queryMods = append(queryMods, qm.GroupBy("np.type_slug"))
+	}
+
```
Comment (fishnix) [inline internal/dbtools/notification_preferences.go:101]: Would it be better to create static raw queries like was done in hierarchical groups for both "withDefaults" and not? This is pretty complicated :laughing:

Comment (fishnix) [reply internal/dbtools/notification_preferences.go:101]: another thought - would it be more straight forward to do the query for the user and if there were no results, return the default?

**Review by fishnix — COMMENTED** [review]
I made some comments inline, this looks like a good start.

File: pkg/api/v1alpha1/users.go [hunk 13]
```diff
@@ -375,6 +410,19 @@ func (r *Router) updateUser(c *gin.Context) {
...
+
 	tx, err := r.DB.BeginTx(c.Request.Context(), nil)
```
Comment (jacobsee) [inline pkg/api/v1alpha1/users.go:426]: How does it work to pass `r.DB` to a separate async function above and then reuse it immediately after to start a transaction? Are the queries made in the async function then wrapped into the transaction as well? This feels like a race condition to me...

File: pkg/events/v1alpha1/events.go [hunk 14]
```diff
@@ -33,16 +33,24 @@ const (
 	GovernorApplicationLinkRequestsEventSubject = "applinks.requests"
 	// GovernorApplicationTypesEventSubject is the subject name for application type events (minus the subject prefix)
 	GovernorApplicationTypesEventSubject = "applicationtypes"
+	// GovernorNotificationTypesEventSubject is the subject name for notification type events (minus the subject prefix)
+	GovernorNotificationTypesEventSubject = "notificationtypes"
+	// GovernorNotificationTargetsEventSubject is the subject name for notification target events (minus the subject prefix)
+	GovernorNotificationTargetsEventSubject = "notificationtarget"
```
Comment (jacobsee) [inline pkg/events/v1alpha1/events.go:39]: slight preference towards plural naming since the others already are

File: sqlboiler.toml (hunk 11 above)
Comment (jacobsee) [reply sqlboiler.toml:12]: because it's a view and the upstream sqlboiler crdb driver crashes on encountering one - but I think if we switched over to https://github.com/metal-toolbox/sqlboiler-crdb now it wouldn't be needed (and sqlboiler would do codegen for it)

**Review by jacobsee — CHANGES_REQUESTED** [review]

File: pkg/api/v1alpha1/notification_preferences.go [hunk 15]
```diff
@@ -0,0 +1,213 @@
...
+	"github.com/metal-toolbox/auditevent/ginaudit"
+	"github.com/metal-toolbox/governor-api/internal/dbtools"
+	"github.com/metal-toolbox/governor-api/internal/eventbus"
+	"github.com/metal-toolbox/governor-api/internal/models"
+	events "github.com/metal-toolbox/governor-api/pkg/events/v1alpha1"
+)
+
+// UserNotificationPreferences is an alias export for the same struct in
+// dbtools
+type UserNotificationPreferences = dbtools.UserNotificationPreferences
```
[the remark this answers is not in the document]
Comment (bailinhe) [reply pkg/api/v1alpha1/notification_preferences.go:20]: something like this?

```go
type UserNotificationPreferences struct{
	dbtools.UserNotificationPreferences
}
```

File: pkg/api/v1alpha1/notification_preferences.go [hunk 16]
```diff
@@ -0,0 +1,213 @@
...
+
+			if err := tx.Rollback(); err != nil {
+				msg = msg + "error rolling back transaction: " + err.Error()
+			}
+
+			return nil, http.StatusBadRequest, newErrUpdateNotificationPreferences(msg)
+		}
+	}
+
+	np, err := dbtools.GetNotificationPreferences(c.Request.Context(), user.ID, db, true)
```
[the remark this answers is not in the document]
Comment (bailinhe) [reply pkg/api/v1alpha1/notification_preferences.go:88]: I defined `handleUpdateNotificationPreferencesRequests` as a helper function to handle all the update related logics. Note that it is not part of the `*Router` object and user validation should be handled by the API handler (see `Router. updateUserNotificationPreferences` and `Router. updateAuthenticatedUserNotificationPreferences`. If it failed to retrieve the preferences after all the lines above executed correctly, I think something really bad would have happened.

File: pkg/api/v1alpha1/router.go (hunk 6 above)
Comment (bailinhe) [reply pkg/api/v1alpha1/router.go:475]: I would go for simplicity and keep it as is until we have a requirement for having two scopes

File: pkg/events/v1alpha1/events.go (hunk 10 above)
Comment (bailinhe) [reply pkg/events/v1alpha1/events.go:41]: would it be a good idea to also use this for `notification-preferences`, e.g. `users.notificationpreferences` or `users.preferences.notifications`?

File: internal/dbtools/notification_preferences.go (hunk 12 above)
Comment (bailinhe) [reply internal/dbtools/notification_preferences.go:101]: yes a raw query looks much better.

we can't do this with the current data model where a user might have only one `type-target` config and the rest is from the `defaults` view

File: internal/dbtools/notification_preferences.go [hunk 17]
```diff
@@ -0,0 +1,303 @@
...
+	}{}
+
+	q := models.NewQuery(
+		qm.Select(
+			"1 as tmp",
+			"jsonb_object_agg(slug, id) as map",
+		),
+		qm.From(table),
+		qm.Where("deleted_at IS NULL"),
+		qm.GroupBy("1"),
```
[the remark this answers is not in the document]
Comment (bailinhe) [reply internal/dbtools/notification_preferences.go]: I was having some brain fog while working on this 😅

File: pkg/api/v1alpha1/notification_preferences.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (fishnix) [reply pkg/api/v1alpha1/notification_preferences.go:19]: yep exactly

File: pkg/api/v1alpha1/users.go [hunk 18]
```diff
@@ -322,7 +362,7 @@ func (r *Router) updateUser(c *gin.Context) {
 		return
 	}
 
-	if req == (UserReq{}) {
+	if req.IsEmpty() {
```
Comment (fishnix) [inline pkg/api/v1alpha1/users.go:365]: I hadn't noticed this before - I'm not a huge fan of adding more reflection although its pretty safe here. Maybe we're better doing the checks below and marking an `updated` bool that we check before starting the transaction?

File: pkg/api/v1alpha1/users.go [hunk 19]
```diff
@@ -418,6 +458,27 @@ func (r *Router) updateUser(c *gin.Context) {
 		return
 	}
 
+	updateNotificationPublishEventErr := error(nil)
```
Comment (fishnix) [inline pkg/api/v1alpha1/users.go:461]:
```suggestion
	var updateNotificationPublishEventErr error
```

Comment (fishnix) [inline pkg/api/v1alpha1/users.go:461]:
```suggestion
	var updateNotificationPublishEventErr error
```

File: pkg/api/v1alpha1/authenticated_user.go [hunk 20]
```diff
@@ -520,6 +527,27 @@ func (r *Router) updateAuthenticatedUser(c *gin.Context) {
 		return
 	}
 
+	updateNotificationPublishEventErr := error(nil)
```
Comment (fishnix) [inline pkg/api/v1alpha1/authenticated_user.go:530]:
```suggestion
	var updateNotificationPublishEventErr error
```

is this your intention here?

File: pkg/api/v1alpha1/authenticated_user.go [hunk 21]
```diff
@@ -520,6 +527,27 @@ func (r *Router) updateAuthenticatedUser(c *gin.Context) {
...
+
+	_, status, err := handleUpdateNotificationPreferencesRequests(
+		c, tx, ctxUser, r.EventBus, req.NotificationPreferences,
+	)
+	if err != nil && !errors.Is(err, ErrNotificationPreferencesEmptyInput) {
+		if errors.Is(err, ErrPublishUpdateNotificationPreferences) {
```
Comment (fishnix) [inline pkg/api/v1alpha1/authenticated_user.go:536]: just a thought - wouldn't it be simpler to duplicate some of the code for updating notification preferences here and the other places where we're trying to use abstracted code? 😄  i believe other cases are handled that way.

File: pkg/events/v1alpha1/events.go (hunk 10 above)
Comment (fishnix) [reply pkg/events/v1alpha1/events.go]: should this be handled by an update event on user instead? what limitations might that impose on downstream systems?

File: pkg/api/v1alpha1/notification_preferences.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (bailinhe) [reply pkg/api/v1alpha1/notification_preferences.go:19]: yeah I tried that and it didn't work exactly as a type alias because `dbtools.UserNotificationPreferences` itself is an array

File: pkg/api/v1alpha1/users.go (hunk 18 above)
Comment (bailinhe) [reply pkg/api/v1alpha1/users.go:365]: `go` won't let me use `==` to compare `UserNotificationPreferences`, since it's an array

File: pkg/api/v1alpha1/authenticated_user.go (hunk 21 above)
Comment (bailinhe) [reply pkg/api/v1alpha1/authenticated_user.go:536]: I was doing that, until I had to do the same update after I changed the model in four different places, then I decided it should be abstracted 😄

File: pkg/events/v1alpha1/events.go (hunk 10 above)
Comment (bailinhe) [reply pkg/events/v1alpha1/events.go]: arguably none. I implemented this as an effort of following what everything else had done, but I can hardly come up with a reason for the downstream systems to catch a user notification update event, unless we are trying to cache the notification preferences downstream.

It could make sense if we are to support a more sophisticated user preferences, e.g. to support more than just notification preferences, then it might be good to have a `users.preferences.*`

File: pkg/api/v1alpha1/router.go [hunk 22]
```diff
@@ -119,6 +135,20 @@ func (r *Router) Routes(rg *gin.RouterGroup) {
...
+		r.getUserNotificationPreferences,
+	)
+
+	rg.PUT(
+		"/users/:id/notification-preferences",
+		r.AuditMW.AuditWithType("UpdateUserNotificationPreferences"),
+		r.AuthMW.AuthRequired(updateScopesWithOpenID("governor:users")),
+		r.updateUserNotificationPreferences,
+	)
+
```
[the remark this answers is not in the document]
Comment (bailinhe) [reply pkg/api/v1alpha1/router.go:151]: I mean it's there if anyone need it, I also updated the clients to support this.

File: pkg/api/v1alpha1/users.go (hunk 18 above)
Comment (fishnix) [reply pkg/api/v1alpha1/users.go:367]: Sure, but why do we need this check at all?

File: pkg/events/v1alpha1/events.go (hunk 10 above)
Comment (fishnix) [reply pkg/events/v1alpha1/events.go]: yeah I would argue we can implement it at that time then?

File: pkg/api/v1alpha1/router.go (hunk 22 above)
[the remark this answers is not in the document]
Comment (bailinhe) [reply pkg/api/v1alpha1/router.go]: Admin updates on user notifications preferences are removed

File: db/migrations/00031_notification_preferences.sql [hunk 23]
```diff
@@ -0,0 +1,77 @@
...
+CREATE TABLE notification_preferences (
+  id UUID PRIMARY KEY NOT NULL DEFAULT gen_random_uuid(),
+  user_id UUID NOT NULL REFERENCES users(id) ON DELETE CASCADE ON UPDATE CASCADE,
+  notification_type_id UUID NOT NULL REFERENCES notification_types(id) ON DELETE CASCADE ON UPDATE CASCADE, 
+  notification_target_id UUID NULL REFERENCES notification_targets(id) ON DELETE CASCADE ON UPDATE CASCADE,
+  notification_target_id_null_string UUID AS (IFNULL(notification_target_id, '00000000-0000-0000-0000-000000000000')) STORED,
+  enabled BOOL NOT NULL DEFAULT false,
+
+  CONSTRAINT unique_user_type_target UNIQUE (user_id, notification_type_id, notification_target_id_null_string),
+  INDEX (user_id) STORING (notification_type_id, notification_target_id, notification_target_id_null_string, enabled),
```
Comment (msloan-pckt) [inline db/migrations/00031_notification_preferences.sql:40]: there seems to be a lot of overlap here on these two indexes. 


`INDEX (user_id) STORING (notification_type_id, notification_target_id, notification_target_id_null_string, enabled) `

and 
`INDEX (user_id, notification_target_id, notification_type_id) STORING (enabled)`


The crdb query planner still has some interesting quirks to me, however in my head something like the following might work as a single index. 

` INDEX (user_id, notification_target_id, notification_type_id) STORING (enabled,  notification_target_id_null_string)`


However it is possible I'm overlooking something.

File: db/migrations/00031_notification_preferences.sql [hunk 24]
```diff
@@ -0,0 +1,77 @@
...
+CREATE MATERIALIZED VIEW notification_defaults AS 
+SELECT
+  targets.id as target_id,
+  targets.slug as target_slug,
+  types.id as type_id,
+  types.slug as type_slug,
+  targets.default_enabled as default_enabled
+FROM
+  notification_types as types
+CROSS JOIN notification_targets as targets
```
[the remark this answers is not in the document]
Comment (bailinhe) [reply db/migrations/00031_notification_preferences.sql:53]: exactly, effective producing something similar to this
```json
[
	{
		"notification_type": "defcon-1",
		"enabled": true,
		"notification_targets": [
			{
				"target": "slack",
				"enabled": true
			},
			{
				"target": "email",
				"enabled": true
			}
		]
	},
	{
		"notification_type": "notice",
		"enabled": false,
		"notification_targets": [
			{
				"target": "slack",
				"enabled": true
			},
			{
				"target": "email",
				"enabled": true
			}
		]
	},
	{
		"notification_type": "alert",
		"enabled": true,
		"notification_targets": [
			{
				"target": "slack",
				"enabled": true
			},
			{
				"target": "email",
				"enabled": true
			}
		]
	}
]
```

I'd be very happy if there's a more elegant way to do this

File: db/migrations/00031_notification_preferences.sql (hunk 23 above)
Comment (bailinhe) [reply db/migrations/00031_notification_preferences.sql:40]: I ran your index on the [query that created the suggestions for the first two indices](https://github.com/metal-toolbox/governor-api/pull/15/files#diff-4b14d273a00680e0d2e6e7c82e731bb0e45a73768b009676976b3a52b6d86e06R99) with `explain`, everything looks ok. I will commit the suggestion.

File: internal/dbtools/notification_preferences.go [hunk 25]
```diff
@@ -0,0 +1,293 @@
...
+		return nil, err
+	}
+
+	eg := &errgroup.Group{}
+	preferences := make(UserNotificationPreferences, len(records))
+
+	for i, r := range records {
+		i, r := i, r // https://golang.org/doc/faq#closures_and_goroutines
+
+		eg.Go(func() error {
```
Comment (fishnix) [inline internal/dbtools/notification_preferences.go:129]: did you explore how long this takes? looking at that package, it seems like negative limits remove the limt and a positive limit will ensure only x goroutines run at a time, but I'm not sure what the zero value limit is?

File: pkg/api/v1alpha1/router.go [hunk 26]
```diff
@@ -119,6 +135,13 @@ func (r *Router) Routes(rg *gin.RouterGroup) {
 		r.deleteUser,
 	)
 
+	rg.GET(
+		"/users/:id/notification-preferences",
```
Comment (fishnix) [inline pkg/api/v1alpha1/router.go:139]: I thought we were removing these in favor of `/users/:id` ?

**Review by fishnix — APPROVED** [review]
just the two questions, otherwise looks good 👍

**Review by jacobsee — APPROVED** [review]

**Review by fishnix — APPROVED** [review]
